runtime/interpreter/decode.go

```diff
@@ -1551,572 +1447,23 @@ func (d *DecoderV5) decodeRestrictedStaticType() (StaticType, error) {
 	}, nil
 }
 
-func (d *DecoderV5) decodeType() (TypeValue, error) {
-	const expectedLength = encodedTypeValueTypeLength
-
-	arraySize, err := d.decoder.DecodeArrayHead()
-
-	if err != nil {
-		if e, ok := err.(*cbor.WrongTypeError); ok {
-			return TypeValue{}, fmt.Errorf("invalid type encoding: expected [%d]interface{}, got %s",
-				expectedLength,
-				e.ActualType.String(),
-			)
-		}
-		return TypeValue{}, err
-	}
-
-	if arraySize != expectedLength {
-		return TypeValue{}, fmt.Errorf("invalid type encoding: expected [%d]interface{}, got [%d]interface{}",
-			expectedLength,
-			arraySize,
-		)
-	}
-
-	// Decode type at array index encodedTypeValueTypeFieldKeyV5
-	var staticType StaticType
-
-	// Optional type can be CBOR nil.
-	err = d.decoder.DecodeNil()
-	if _, ok := err.(*cbor.WrongTypeError); ok {
-		staticType, err = d.decodeStaticType()
-	}
-
-	if err != nil {
-		return TypeValue{}, fmt.Errorf("invalid type encoding: %w", err)
-	}
-
-	return TypeValue{
-		Type: staticType,
-	}, nil
-}
-
-func (d *DecoderV5) decodeCapabilityStaticType() (StaticType, error) {
+func decodeCapabilityStaticType(dec *cbor.StreamDecoder) (StaticType, error) {
 	var borrowStaticType StaticType
 
 	// Optional borrow type can be CBOR nil.
-	err := d.decoder.DecodeNil()
+	err := dec.DecodeNil()
 	if _, ok := err.(*cbor.WrongTypeError); ok {
-		borrowStaticType, err = d.decodeStaticType()
+		borrowStaticType, err = decodeStaticType(dec)
 	}
 
 	if err != nil {
-		return nil, fmt.Errorf("invalid capability static type borrow type encoding: %w", err)
+		return nil, fmt.Errorf(
+			"invalid capability static type borrow type encoding: %w",
+			err,
+		)
 	}
 
 	return CapabilityStaticType{
 		BorrowType: borrowStaticType,
 	}, nil
 }
```

Comment:
This removed code was an optimization we had added earlier, and is now not applicable anymore: Lazy decoding of stored composites, dictionaries, and arrays, until they are actually accessed